tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java

```diff
@@ -0,0 +1,227 @@
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.gson.Gson;
+import com.uber.tchannel.api.Request;
+import com.uber.tchannel.api.Response;
+import com.uber.tchannel.api.TChannel;
+import com.uber.tchannel.api.errors.TChannelError;
+import com.uber.tchannel.hyperbahn.messages.AdvertiseRequest;
+import com.uber.tchannel.hyperbahn.messages.AdvertiseResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class HyperbahnClient {
+    private static final String HYPERBAHN_SERVICE_NAME = "hyperbahn";
+    private static final String HYPERBAHN_ADVERTISE_ENDPOINT = "ad";
+    public AtomicBoolean destroyed = new AtomicBoolean(false);
+    private final Logger logger = LoggerFactory.getLogger(HyperbahnClient.class);
+
+    private final String service;
+    private final TChannel tchannel;
+    private final List<InetSocketAddress> routers;
+    private final long advertiseTimeout;
+    private final long advertiseInterval;
+
+    private Timer advertiseTimer = new Timer(true);
+
+    private HyperbahnClient(Builder builder) {
+        this.service = builder.service;
+        this.tchannel = builder.channel;
+        this.routers = builder.routers;
+        this.advertiseTimeout = builder.advertiseTimeout;
+        this.advertiseInterval = builder.advertiseInterval;
+    }
+
+    public ListenableFuture<Response<AdvertiseResponse>> advertise()
+        throws InterruptedException, TChannelError {
+
+        final AdvertiseRequest advertiseRequest = new AdvertiseRequest();
+        advertiseRequest.addService(service, 0);
+
+        // TODO: options for timeout, hard fail, etc.
+        final Request<AdvertiseRequest> request = new Request.Builder<>(
+            advertiseRequest,
+            HYPERBAHN_SERVICE_NAME,
+            HYPERBAHN_ADVERTISE_ENDPOINT
+        )
+            .setTTL(advertiseTimeout, TimeUnit.SECONDS)
+            .build();
+
+        // TODO: should be part of tchannel peer selection
+        final InetSocketAddress router = this.routers.get(new Random().nextInt(this.routers.size()));
+        ListenableFuture<Response<AdvertiseResponse>> responseFuture = null;
+        try {
+            responseFuture = this.tchannel.callJSON(
+                    router.getAddress(),
+                    router.getPort(),
+                    request,
+                    AdvertiseResponse.class
+            );
+        } catch (TChannelError ex) {
+            // TODO: should be moved into tchanne as retry ...
+            this.logger.error("Advertise failure: " + ex.toString());
+
+            // re-throw for now
+            // TODO: should really handle internally
+            throw ex;
+        }
+
+        responseFuture.addListener(new Runnable() {
+            @Override
+            public void run() {
+                if (destroyed.get()) {
+                    return;
+                }
+
+                scheduleAdvertise();
+            }
+        }, new Executor() {
```

Comment:
Doesn't really matter when this ran once a minute.
Feel free to suggest a better style. I can make the update. Or you can send a PR for it. :)